File: step1/2.3_dynamic_programming/53_maxSubArray.cpp
#include <iostream>
#include <vector>
#include <limits.h>

using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int f = 0;
        int result = INT_MIN;
        for(int i = 0; i< nums.size(); i++)
        {
            int A,B;
            f = max(f + nums[i], nums[i]);
            result = max(result,f);
        }
        return result;
    }
};

int main(){
    Solution* sln = new Solution();
    vector<int> nums({-2,1,-3,4,-1,2,1,-5,4});
    int res;
    res = sln->maxSubArray(nums);
    cout<<res<<endl;
    return 0;
}

File: step1/2.3_dynamic_programming/53_maxSubArray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        f = [0]*len(nums)
        for i in range(0,len(nums)):

            if i == 0:
                f[i] = nums[i]
            else:
                A = f[i-1] + nums[i]
                B = nums[i]
                f[i] = max(A,B)
            
        return max(f)


nums = [-2,1,-3,4,-1,2,1,-5,4]
sln = Solution()
res = sln.maxSubArray(nums)
print(res)
